File: app/src/main/java/com/wa/pranksound/activity/SoundDetailActivity.java
package com.wa.pranksound.activity;

import static com.wa.pranksound.utils.KeyClass.cate_name;
import static com.wa.pranksound.utils.KeyClass.image_sound;
import static com.wa.pranksound.utils.KeyClass.is_fav;
import static com.wa.pranksound.utils.KeyClass.music_name;
import static com.wa.pranksound.utils.KeyClass.sound_path;

import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import android.annotation.SuppressLint;
import android.content.res.AssetFileDescriptor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.text.format.DateUtils;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Switch;
import android.widget.TextView;

import com.airbnb.lottie.LottieAnimationView;
import com.bumptech.glide.Glide;
import com.wa.pranksound.R;
import com.wa.pranksound.adapter.CategoriesAdapter;
import com.wa.pranksound.adapter.HorizontalFavoriteSoundAdapter;
import com.wa.pranksound.Room.AppDatabase;
import com.wa.pranksound.Room.InsertPrankSound;
import com.wa.pranksound.Room.QueryClass;
import com.wa.pranksound.utils.BaseActivity;
import com.wa.pranksound.utils.Utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class SoundDetailActivity extends BaseActivity {
    ImageView imgBack, imgItem, imgPlayPause, imgHeart;
    LottieAnimationView imgAnimation;

    Boolean isPlaying = false, isLoop = false;
    String strMusicName;
    TextView txtTitle, endTime, txtCountTime, tvOff;
    Vibrator systemService;
    MediaPlayer mediaPlayer = new MediaPlayer();
    ProgressBar seekBar;
    Handler handler = new Handler();
    Runnable runnable;
    Switch swLoop;
    RecyclerView rvSound;
    QueryClass queryClass;
    String strCateName, musicName;
    List<InsertPrankSound> arrFavPrankSound = new ArrayList<>();
    HorizontalFavoriteSoundAdapter horizontalSoundAdapter;
    Boolean isFav = false;
    LinearLayout llBtnOff;
    CountDownTimer countDownTimer;
    List<String> cate;
    String[] images;
    String string_img_sound;
    Integer int_img_sound;
    String soundName;
    String soundPath;

    FrameLayout fl_banner;
    View view_line;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sound_detail);
        AppDatabase db = Room.databaseBuilder(this, AppDatabase.class, "prank_sound").allowMainThreadQueries().fallbackToDestructiveMigration().build();
        queryClass = db.queryClass();

        findView();
        clickEvent();
    }

    private void findView() {
        imgBack = findViewById(R.id.imgBack);
        imgItem = findViewById(R.id.imgItem);
        txtTitle = findViewById(R.id.txtTitle);
        imgPlayPause = findViewById(R.id.imgPlayPause);
        endTime = findViewById(R.id.endTime);
        seekBar = findViewById(R.id.seekBar);
        swLoop = findViewById(R.id.swLoop);
        rvSound = findViewById(R.id.rvSound);
        imgHeart = findViewById(R.id.imgHeart);
        llBtnOff = findViewById(R.id.llBtnOff);
        txtCountTime = findViewById(R.id.txtCountTime);
        tvOff = findViewById(R.id.tvOff);

        imgAnimation = findViewById(R.id.animation);

        view_line = findViewById(R.id.line);

        systemService = (Vibrator) getSystemService(VIBRATOR_SERVICE);

        strMusicName = getIntent().getStringExtra(music_name);

        isFav = getIntent().getBooleanExtra(is_fav, false);
        strCateName = getIntent().getStringExtra(cate_name);

        //get music name from sound list
        musicName = getIntent().getStringExtra(music_name);
        string_img_sound = getIntent().getStringExtra(image_sound);
        int_img_sound = getIntent().getIntExtra(image_sound, 0);

        soundPath = getIntent().getStringExtra(sound_path);

        Log.d("check_file", "sound path: " + soundPath);
        Log.d("check_file", "image_sound_in_detail: " + int_img_sound);
        Log.d("check_file", "image_sound_from_fav: " + string_img_sound);
        Log.d("check_file", "music name: " + musicName);

        InsertPrankSound insertPrankSound1 = queryClass.getFavSound(strCateName, strMusicName);
        if (insertPrankSound1 != null) {
            imgHeart.setImageResource(R.drawable.ic_favorite_check);
        } else {
            imgHeart.setImageResource(R.drawable.ic_heart);
        }

        if (strMusicName != null) {
            txtTitle.setText(musicName);
        }
        Log.d("check_file", "str cate name: " + strCateName);
        if (strCateName != null) {
            try {
                cate = Arrays.asList(Objects.requireNonNull(getAssets().list("prank_sound")));
                images = getAssets().list("prank_sound/" + strCateName);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }

            CategoriesAdapter categoriesAdapter = new CategoriesAdapter(this, cate);
            LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
            rvSound.setLayoutManager(layoutManager);
            rvSound.setAdapter(categoriesAdapter);

            //load ảnh từ file asset
            if (int_img_sound != 0) {
                Bitmap bitmap = BitmapFactory.decodeResource(getResources(), int_img_sound);
                Glide.with(this).load(bitmap).into(imgItem);
            } else {
                if (!string_img_sound.contains("png")) {
                    Bitmap bitmap = BitmapFactory.decodeResource(getResources(), Integer.parseInt(string_img_sound));
                    Glide.with(this).load(bitmap).into(imgItem);
                } else {
                    Glide.with(this).load("file:///android_asset/" + string_img_sound).into(imgItem);
                }
            }

            try {
                Log.d("check_file", "str cate name in sound: " + strCateName);
                if (strCateName.equals("record")) {
                    soundPath = musicName;
                }
                //set sound from asset or record
                if (soundPath != null) {
                    try {
                        // Đặt nguồn dữ liệu cho MediaPlayer từ File
                        mediaPlayer.setDataSource(soundPath);
                    } catch (IOException ignored) {
                    }
                } else {
                    AssetFileDescriptor descriptor;
                    descriptor = getAssets().openFd("prank_sound/" + strCateName + "/" + strMusicName + ".ogg");
                    Log.d("check_file", "des path: " + "prank_sound/" + strCateName + "/" + strMusicName + ".ogg");
                    Log.d("check_file", "descriptor: " + descriptor);
                    mediaPlayer.setDataSource(descriptor.getFileDescriptor(), descriptor.getStartOffset(), descriptor.getLength());
                    descriptor.close();

                }
                mediaPlayer.prepare();
                Log.d("check_file", "str cate name outttt: " + strCateName);
                int length = mediaPlayer.getDuration(); // lấy giá trị độ dài của sound

                String durationText = DateUtils.formatElapsedTime(length / 1000); // converting time in millis to minutes:second format eg 14:15 min
                endTime.setText(durationText);
                seekBar.setMax(length);

                runnable = () -> {
                    if (mediaPlayer != null) {
                        seekBar.setProgress(mediaPlayer.getCurrentPosition());
                        handler.postDelayed(runnable, 5);
                    }

                };

                mediaPlayer.setOnCompletionListener(mp -> {
                    if (!isLoop) {
                        if (mediaPlayer != null) {
                            seekBar.setProgress(mediaPlayer.getDuration());
                        }
                        isPlaying = false;
                        imgPlayPause.setImageResource(R.drawable.play);
                        //set anim
                        imgAnimation.setVisibility(View.INVISIBLE);
                    } else {
                        seekBar.setProgress(0);
                        mediaPlayer.start();
                        startVibrate();
                    }
                });

            } catch (Exception e) {
                Log.d("check_file", "error " + e);
                e.printStackTrace();
            }

            swLoop.setOnCheckedChangeListener((buttonView, isChecked) -> isLoop = isChecked);
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private void clickEvent() {
        imgBack.setOnClickListener(v -> {
            Log.d("check_show_ads", "show in back sound list");
            getOnBackPressedDispatcher().onBackPressed();
        });
        llBtnOff.setOnClickListener(v -> {
            PopupMenu popupMenu = new PopupMenu(new ContextThemeWrapper(SoundDetailActivity.this, R.style.myPopupMenu), v);
            popupMenu.getMenuInflater().inflate(R.menu.set_time, popupMenu.getMenu());
            popupMenu.setOnMenuItemClickListener(item -> {
                if (item.getItemId() == R.id.five) {
                    startCountDownTimer(5000, txtCountTime);
                    tvOff.setText(R.string._5s);
                    return true;
                } else if (item.getItemId() == R.id.fiveMinute) {
                    startCountDownTimer(300000, txtCountTime);
                    tvOff.setText(R.string._5m);
                    return true;
                } else if (item.getItemId() == R.id.off) {
                    if (countDownTimer != null) {
                        countDownTimer.cancel();
                    }
                    imgPlayPause.setEnabled(true);
                    txtCountTime.setText("");
                    tvOff.setText(R.string._off);
                    return true;
                } else if (item.getItemId() == R.id.oneMinute) {
                    startCountDownTimer(60000, txtCountTime);
                    tvOff.setText(R.string._1m);
                    return true;
                } else if (item.getItemId() == R.id.ten) {
                    startCountDownTimer(10000, txtCountTime);
                    tvOff.setText(R.string._10s);
                    return true;
                } else if (item.getItemId() == R.id.thirty) {
                    startCountDownTimer(30000, txtCountTime);
                    tvOff.setText(R.string._30s);
                    return true;
                } else {
                    return false;
                }
            });
            popupMenu.show();
        });

        //set playing when click item
        imgItem.setOnTouchListener((v, event) -> {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    if (mediaPlayer != null) {
                        if (!mediaPlayer.isPlaying()) {
                            imgAnimation.setVisibility(View.VISIBLE);
                            imgAnimation.playAnimation();
                            mediaPlayer.seekTo(0);
                            mediaPlayer.setLooping(true);
                            mediaPlayer.start();
                            startVibrate();
                        } else {
                            if (isLoop) {
                                mediaPlayer.pause();
                                stopVibrate();
                                imgAnimation.setVisibility(View.INVISIBLE);
                            }
                        }
                    }
                    return true;
                case MotionEvent.ACTION_UP:
                    if (!isLoop) {
                        if (mediaPlayer != null) {
                            if (mediaPlayer.isPlaying()) {
                                mediaPlayer.pause();
                                stopVibrate();
                                imgAnimation.setVisibility(View.INVISIBLE);
                            }
                        }
                    }
                    return false;
            }
            return false;
        });

        imgPlayPause.setOnClickListener(v -> {
            if (isPlaying) {
                isPlaying = false;
                imgPlayPause.setImageResource(R.drawable.play);
                if (mediaPlayer != null) {
                    if (mediaPlayer.isPlaying()) {
                        mediaPlayer.pause();
                        stopVibrate();
                    }
                }
            } else {
                isPlaying = true;
                imgPlayPause.setImageResource(R.drawable.pause);
                if (mediaPlayer != null) {
                    if (!mediaPlayer.isPlaying()) {
                        mediaPlayer.start();
                        startVibrate();
                        handler.postDelayed(runnable, 5);
                    }

                }

            }

        });


        //test favories
        imgHeart.setOnClickListener(v -> {
            InsertPrankSound insertPrankSound1 = queryClass.getFavSound(strCateName, strMusicName);
            if (insertPrankSound1 != null) {
                queryClass.getUnFavSound(strCateName, strMusicName);
                imgHeart.setImageResource(R.drawable.ic_heart);
            } else {
                InsertPrankSound insertPrankSound = new InsertPrankSound();
                insertPrankSound.folder_name = strCateName;
                insertPrankSound.sound_name = strMusicName;

                if (int_img_sound != 0) {
                    insertPrankSound.image_path = int_img_sound.toString();
                    insertPrankSound.sound_path = soundPath;
                } else {
                    insertPrankSound.image_path = string_img_sound;
                }
                Log.d("Img_path", insertPrankSound.image_path);
                queryClass.insertPrankSound(insertPrankSound);
                imgHeart.setImageResource(R.drawable.ic_favorite_check);
            }
            if (isFav) {
                arrFavPrankSound = queryClass.getAllFavSound();
                if (horizontalSoundAdapter != null) {
                    Collections.reverse(arrFavPrankSound);
                    horizontalSoundAdapter.notifyDataSetChangedAd(arrFavPrankSound);
                }
            }
        });
    }

    private void startCountDownTimer(int i, TextView txtCountTime) {
        imgPlayPause.setEnabled(false);
        imgAnimation.setVisibility(View.INVISIBLE);
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        countDownTimer = new CountDownTimer(i, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                long j2 = millisUntilFinished / ((long) 1000);
                long j3 = (long) 60;
                long j4 = j2 / j3;
                long j5 = j2 % j3;
                String format = String.format("%02d:%02d", Arrays.copyOf(new Object[]{Long.valueOf(j4), Long.valueOf(j5)}, 2));
                txtCountTime.setText(format);
            }

            @Override
            public void onFinish() {
                tvOff.setText(R.string._off);
                txtCountTime.setText("");
                imgPlayPause.setEnabled(true);
                isPlaying = true;
                imgPlayPause.setImageResource(R.drawable.pause);
                imgAnimation.setVisibility(View.VISIBLE);
                if (mediaPlayer != null) {
                    if (!mediaPlayer.isPlaying()) {
                        mediaPlayer.start();
                        startVibrate();
                        handler.postDelayed(runnable, 5);
                    }
                }
            }
        }.start();

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mediaPlayer != null) {
            if (mediaPlayer.isPlaying()) {
                stopVibrate();
                mediaPlayer.reset();
                mediaPlayer.stop();
                mediaPlayer.release();
                mediaPlayer = null;

            }
        }
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        handler.removeCallbacksAndMessages(null);
        handler.removeCallbacks(runnable);
    }

    @Override
    protected void onStop() {
        super.onStop();

    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mediaPlayer != null) {
            if (mediaPlayer.isPlaying()) {
                stopVibrate();
                mediaPlayer.reset();
                mediaPlayer.stop();
                mediaPlayer.release();
                mediaPlayer = null;

            }
        }
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
        handler.removeCallbacksAndMessages(null);
        handler.removeCallbacks(runnable);
    }

    public final void startVibrate() {
        if (Utils.INSTANCE.getVibration(this)) {
            if (systemService != null) {
                if (Build.VERSION.SDK_INT < 26 || !systemService.hasVibrator()) {
                    systemService.vibrate((long) mediaPlayer.getDuration());
                } else {
                    systemService.vibrate(VibrationEffect.createWaveform(new long[]{0, 1000}, 0));
                }
            }
        }
    }

    public final void stopVibrate() {
        if (systemService != null) {
            systemService.cancel();

        }
    }
}

File: app/src/main/java/com/joya/pranksound/ui/component/sound/SoundDetailActivity.kt
package com.joya.pranksound.ui.component.sound

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.PopupWindow
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.room.Room.databaseBuilder
import com.adjust.sdk.Adjust
import com.bumptech.glide.Glide
import com.joya.pranksound.R
import com.joya.pranksound.adapter.HorizontalFavoriteSoundAdapter
import com.joya.pranksound.adapter.HorizontalSoundAdapterTest
import com.joya.pranksound.common.Constant.KEY_IS_FIRST_OPEN_SOUND
import com.joya.pranksound.data.SharedPreferenceHelper.Companion.getBoolean
import com.joya.pranksound.data.SharedPreferenceHelper.Companion.storeBoolean
import com.joya.pranksound.databinding.ActivitySoundDetailBinding
import com.joya.pranksound.databinding.MenuTimerBinding
import com.joya.pranksound.model.Sound
import com.joya.pranksound.room.AppDatabase
import com.joya.pranksound.room.InsertPrankSound
import com.joya.pranksound.room.QueryClass
import com.joya.pranksound.ui.base.BaseBindingActivity
import com.joya.pranksound.ui.component.main.MainActivity
import com.joya.pranksound.utils.ImageLoader
import com.joya.pranksound.utils.KeyClass
import com.joya.pranksound.utils.Utils.getAudioList
import com.joya.pranksound.utils.Utils.getVibration
import com.joya.pranksound.utils.Utils.removeAfterDot
import com.joya.pranksound.utils.Utils.saveAudioList
import com.joya.pranksound.utils.extention.gone
import com.joya.pranksound.utils.extention.invisible
import com.joya.pranksound.utils.extention.setOnSafeClick
import com.joya.pranksound.utils.extention.visible
import java.io.IOException
import java.util.Objects

class SoundDetailActivity :
    BaseBindingActivity<ActivitySoundDetailBinding, SoundDetailViewModel>() {

    var isPlaying: Boolean = false
    private var isLoop: Boolean = false
    private var strMusicName: String? = null
    private var systemService: Vibrator? = null
    var mediaPlayer: MediaPlayer? = MediaPlayer()

    var handler: Handler = Handler(Looper.getMainLooper())
    var runnable: Runnable? = null
    private var queryClass: QueryClass? = null
    private var strCateName: String? = null
    private var arrFavPrankSound: List<InsertPrankSound?> = ArrayList()
    private var horizontalSoundAdapter: HorizontalFavoriteSoundAdapter? = null
    private var isFav: Boolean = false
    private var countDownTimer: CountDownTimer? = null
    private var cate: List<String>? = null
    private var images: Array<String>? = null
    private var string_img_sound: String? = null
    private var int_img_sound: Int? = null
    private var soundName: String? = null
    private var soundPath: String? = null

    private var audioManager: AudioManager? = null
    private var isRecord: Boolean = false

    override val layoutId: Int
        get() = R.layout.activity_sound_detail

    override fun getViewModel(): Class<SoundDetailViewModel> = SoundDetailViewModel::class.java
    override fun setupView(savedInstanceState: Bundle?) {
        val db =
            databaseBuilder(this, AppDatabase::class.java, "prank_sound").allowMainThreadQueries()
                .fallbackToDestructiveMigration().build()
        queryClass = db.queryClass()
        data
        clickEvent()

        setUpVolume()
    }

    override fun setupData() {
    }

    private fun setUpVolume() {
        audioManager = getSystemService(AUDIO_SERVICE) as AudioManager

        val maxVolume = audioManager!!.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        val currentVolume = audioManager!!.getStreamVolume(AudioManager.STREAM_MUSIC)

        binding.volumeSeekBar.max = maxVolume
        binding.volumeSeekBar.progress = currentVolume
        binding.btnVolumeSmall.setOnClickListener {
            binding.volumeSeekBar.progress = maxVolume / 5
        }
        binding.btnVolumeLoud.setOnClickListener {
            binding.volumeSeekBar.progress = 4 * maxVolume / 5
        }
        binding.volumeSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                audioManager!!.setStreamVolume(AudioManager.STREAM_MUSIC, progress, 0)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
            }
        })
    }

    private val data: Unit
        get() {
            val intent = intent
            soundName = intent.getStringExtra("sound_name")

            systemService = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val vibratorManager =
                    getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
                vibratorManager.defaultVibrator
            } else {
                getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            }

            strMusicName = getIntent().getStringExtra(KeyClass.music_name)

            isFav = getIntent().getBooleanExtra(KeyClass.is_fav, false)
            strCateName = getIntent().getStringExtra(KeyClass.cate_name)

            //get music name from sound list
            string_img_sound = getIntent().getStringExtra(KeyClass.image_sound)
            int_img_sound = getIntent().getIntExtra(KeyClass.image_sound, 0)
            isRecord = getIntent().getBooleanExtra(KeyClass.is_record, false)

            soundPath = getIntent().getStringExtra(KeyClass.sound_path)

            setUpSound()

            try {
                cate = listOf(
                    *Objects.requireNonNull(
                        assets.list("prank_sound")
                    )
                )
                images = assets.list("prank_sound/$strCateName")

                val images = assets.list("prank_sound/$strCateName")
                val listSound: MutableList<Sound> = ArrayList()
                checkNotNull(images)

                val imageList = ImageLoader.getImageListFromAssets(this, "prank_image/$strCateName")

                for (i in images.indices) {
                    var sound: Sound
                    val imageName = removeAfterDot(images[i])
                    sound = Sound(strCateName!!, imageList[i], imageName, 0, true, isNew = false)
                    listSound.add(sound)
                }
                val verticalSoundAdapter = HorizontalSoundAdapterTest(listSound) { position: Int ->
                    val sound = listSound[position]
                    getData(sound)
                }
                binding.rvSound.adapter = verticalSoundAdapter
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

    private fun getData(sound: Sound?) {
        if (sound != null) {
            soundName = sound.name
            strMusicName = sound.name
            strCateName = sound.typeSound
            string_img_sound = sound.image
            int_img_sound = 0
            isRecord = false
            soundPath = null
            setUpSound()
        }
    }

    private fun setUpSound() {
        val insertPrankSound1 = queryClass!!.getFavSound(strCateName, strMusicName)
        if (isRecord) {
            binding.llMoreSounds.gone()
            binding.btnDelete.visible()
            binding.imgHeart.gone()
        } else {
            binding.btnDelete.gone()
            binding.imgHeart.visible()
            binding.imgHeart.isSelected = insertPrankSound1 != null
        }


        if (strMusicName != null) {
            binding.tvTitle.text = strMusicName
        }

        if (strCateName != null) {
            //load ảnh từ file asset

            if (int_img_sound != 0) {
                Glide.with(this).load(int_img_sound).into(binding.imgItem)
            } else {
                if (!string_img_sound!!.contains("png")) {
                    val bitmap = BitmapFactory.decodeResource(resources, string_img_sound!!.toInt())
                    Glide.with(this).load(bitmap).into(binding.imgItem)
                } else {
                    Glide.with(this).load("file:///android_asset/$string_img_sound").into(
                        binding.imgItem
                    )
                }
            }

            try {
                if (mediaPlayer != null) {
                    mediaPlayer!!.release()
                    stopVibrate()
                    binding.animation.invisible()
                }
                mediaPlayer = MediaPlayer()
                if (strCateName == "record") {
                    soundPath = strMusicName
                }
                //set sound from asset or record
                if (soundPath != null) {
                    try {
                        mediaPlayer!!.setDataSource(soundPath)
                    } catch (ignored: IOException) {
                    }
                } else {
                    val descriptor = assets.openFd("prank_sound/$strCateName/$strMusicName.ogg")
                    mediaPlayer!!.setDataSource(
                        descriptor.fileDescriptor,
                        descriptor.startOffset,
                        descriptor.length
                    )
                    descriptor.close()
                }
                mediaPlayer!!.prepare()
                runnable = Runnable {
                    if (mediaPlayer != null) {
                        handler.postDelayed(runnable!!, 5)
                    }
                }

                mediaPlayer!!.setOnCompletionListener {
                    if (!isLoop) {
                        isPlaying = false
                        //set anim
                        binding.animation.invisible()
                        stopVibrate()
                    } else {

                        mediaPlayer!!.start()
                        startVibrate()
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun startCountDownTimer(i: Int, txtCountTime: TextView?) {
        binding.animation.visibility = View.INVISIBLE
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
        }
        countDownTimer = object : CountDownTimer(i.toLong(), 1000) {
            override fun onTick(millisUntilFinished: Long) {
                val j2 = millisUntilFinished / (1000L)
                val j3 = 60L
                val j4 = j2 / j3
                val j5 = j2 % j3
                @SuppressLint("DefaultLocale") val format =
                    String.format("%02d:%02d", *arrayOf<Any>(j4, j5).copyOf(2))
                txtCountTime!!.text = format
            }

            override fun onFinish() {
                binding.tvOff.setText(R.string._off)
                val text = ""
                txtCountTime!!.text = text
                isPlaying = true
                binding.animation.visibility = View.VISIBLE
                if (mediaPlayer != null) {
                    if (!mediaPlayer!!.isPlaying) {
                        mediaPlayer!!.start()
                        mediaPlayer!!.isLooping = false
                        startVibrate()
                        handler.postDelayed(runnable!!, 5)
                    }
                }
            }
        }.start()
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun clickEvent() {
        val isFirsTime = getBoolean(KEY_IS_FIRST_OPEN_SOUND, true)
        if (isFirsTime) {
            binding.animGuide.visible()
            binding.viewBlur.visible()
        } else {
            binding.animGuide.gone()
            binding.viewBlur.gone()
        }

        binding.animGuide.setOnTouchListener { _: View?, _: MotionEvent? ->
            storeBoolean(KEY_IS_FIRST_OPEN_SOUND, false)
            binding.animGuide.gone()
            binding.viewBlur.gone()
            true
        }

        binding.btnBack.setOnSafeClick {
            finish()
        }

        binding.llBtnOff.setOnSafeClick { v: View? ->
            val view = MenuTimerBinding.inflate(layoutInflater)
            val popupWindow = PopupWindow(
                view.root,
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )

            view.btnOff.setOnSafeClick {
                if (countDownTimer != null) {
                    countDownTimer!!.cancel()
                }
                val text = ""
                binding.tvCountTime.text = text
                binding.tvOff.setText(R.string._off)
                popupWindow.dismiss()
            }
            view.btn5s.setOnSafeClick {
                startCountDownTimer(5000, binding.tvCountTime)
                binding.tvOff.setText(R.string._5s)
                popupWindow.dismiss()
            }
            view.btn10s.setOnSafeClick {
                startCountDownTimer(10000, binding.tvCountTime)
                binding.tvOff.setText(R.string._10s)
                popupWindow.dismiss()
            }
            view.btn30s.setOnSafeClick {
                startCountDownTimer(30000, binding.tvCountTime)
                binding.tvOff.setText(R.string._30s)
                popupWindow.dismiss()
            }
            view.btn1m.setOnSafeClick {
                startCountDownTimer(60000, binding.tvCountTime)
                binding.tvOff.setText(R.string._1m)
                popupWindow.dismiss()
            }
            view.btn5m.setOnSafeClick {
                startCountDownTimer(300000, binding.tvCountTime)
                binding.tvOff.setText(R.string._5m)
                popupWindow.dismiss()
            }

            popupWindow.isOutsideTouchable = true
            popupWindow.isFocusable = true
            popupWindow.setBackgroundDrawable(null)
            popupWindow.showAsDropDown(v)
        }

        binding.imgItem.setOnTouchListener { _: View?, event: MotionEvent ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    if (mediaPlayer != null) {
                        if (!mediaPlayer!!.isPlaying) {
                            binding.animation.visibility = View.VISIBLE
                            binding.animation.playAnimation()
                            mediaPlayer!!.seekTo(0)
                            mediaPlayer!!.isLooping = true
                            mediaPlayer!!.start()
                            startVibrate()
                            val params = Bundle()
                            if (isRecord) params.putString("Sound_name", "record")
                            else params.putString("Sound_name", soundName)
                        } else {
                            if (isLoop) {
                                mediaPlayer!!.pause()
                                stopVibrate()
                                binding.animation.visibility = View.INVISIBLE
                            }
                        }
                    }
                    true
                }

                MotionEvent.ACTION_UP -> {
                    if (!isLoop) {
                        if (mediaPlayer != null) {
                            if (mediaPlayer!!.isPlaying) {
                                mediaPlayer!!.pause()
                                stopVibrate()
                                binding.animation.visibility = View.INVISIBLE
                            }
                        }
                    }
                    false
                }

                else -> false
            }
        }

        //test favourite
        binding.imgHeart.setOnClickListener {
            val insertPrankSound1 = queryClass!!.getFavSound(strCateName, strMusicName)
            if (insertPrankSound1 != null) {
                queryClass!!.getUnFavSound(strCateName, strMusicName)
                binding.imgHeart.isSelected = false
            } else {
                val insertPrankSound = InsertPrankSound()
                insertPrankSound.folder_name = strCateName
                insertPrankSound.sound_name = strMusicName

                if (int_img_sound != 0) {
                    insertPrankSound.image_path = int_img_sound.toString()
                    insertPrankSound.sound_path = soundPath
                } else {
                    insertPrankSound.image_path = string_img_sound
                }
                queryClass!!.insertPrankSound(insertPrankSound)
                binding.imgHeart.isSelected = true
            }
            if (isFav) {
                arrFavPrankSound = queryClass!!.allFavSound
                if (horizontalSoundAdapter != null) {
                    arrFavPrankSound.reversed()
                    horizontalSoundAdapter!!.notifyDataSetChangedAd(arrFavPrankSound)
                }
            }
        }

        binding.swLoop.setOnClickListener {
            binding.swLoop.isSelected = !binding.swLoop.isSelected
            isLoop = binding.swLoop.isSelected
        }

        binding.btnDelete.setOnClickListener {
            showDiscardDialog {
                deleteRecord(strMusicName)
                startActivity(Intent(this@SoundDetailActivity, MainActivity::class.java))
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mediaPlayer != null) {
            if (mediaPlayer!!.isPlaying) {
                stopVibrate()
                mediaPlayer!!.reset()
                mediaPlayer!!.stop()
                mediaPlayer!!.release()
                mediaPlayer = null
            }
        }
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
        }
        handler.removeCallbacksAndMessages(null)
        handler.removeCallbacks(runnable!!)
    }


    override fun onPause() {
        super.onPause()
        if (mediaPlayer != null) {
            if (mediaPlayer!!.isPlaying) {
                mediaPlayer!!.pause()
                stopVibrate()
                binding.animation.visibility = View.INVISIBLE
            }
        }
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
        }
        handler.removeCallbacksAndMessages(null)
        handler.removeCallbacks(runnable!!)

        Adjust.onPause()
    }

    override fun onResume() {
        super.onResume()
        Adjust.onResume()
    }

    fun startVibrate() {
        if (getVibration(this)) {
            if (systemService != null && systemService!!.hasVibrator()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    val vibrationEffect = VibrationEffect.createWaveform(longArrayOf(0, 1000), 0)
                    systemService!!.vibrate(vibrationEffect)
                } else {
                    systemService!!.vibrate(mediaPlayer!!.duration.toLong())
                }
            }
        }
    }


    private fun stopVibrate() {
        if (systemService != null) {
            systemService!!.cancel()
        }
    }

    private fun deleteRecord(fileName: String?) {
        val audioFileList = getAudioList(this).toMutableList()
        for (i in audioFileList.indices) {
            if (audioFileList[i].name == fileName) {
                audioFileList.removeAt(i)
                break
            }
        }
        saveAudioList(this, audioFileList)
    }

    private fun showDiscardDialog(callback: Runnable) {
        val dialogCustomExit = Dialog(this@SoundDetailActivity)
        dialogCustomExit.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialogCustomExit.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))

        val lp = WindowManager.LayoutParams()
        lp.copyFrom(dialogCustomExit.window!!.attributes)
        lp.width = WindowManager.LayoutParams.MATCH_PARENT
        lp.height = WindowManager.LayoutParams.MATCH_PARENT

        dialogCustomExit.setContentView(R.layout.dialog_delete_record)
        dialogCustomExit.setCancelable(true)
        dialogCustomExit.show()
        dialogCustomExit.window!!.attributes = lp

        val btnNegative = dialogCustomExit.findViewById<TextView>(R.id.btnNegative)
        val btnPositive = dialogCustomExit.findViewById<TextView>(R.id.btnPositive)

        btnPositive.setOnClickListener {
            dialogCustomExit.dismiss()
            callback.run()
        }

        btnNegative.setOnClickListener { dialogCustomExit.dismiss() }
    }
}